internal/build/buildconfig/manager.go

```diff
@@ -0,0 +1,187 @@
+	logger := log.FromContext(ctx)
+
+	config := bcm.helper.GetRelevantBuild(mod, m)
+
+	var registryAuthGetter auth.RegistryAuthGetter
+
+	if irs := mod.Spec.ImageRepoSecret; irs != nil {
+		logger.Info("Getting pull secret from Kubernetes secret", "secret name", irs.Name)
+
+		namespacedName := types.NamespacedName{
+			Name:      irs.Name,
+			Namespace: mod.Namespace,
+		}
+
+		registryAuthGetter = bcm.authFactory.NewRegistryAuthGetter(bcm.client, namespacedName)
+	} else {
+		logger.Info("Getting pull secret from Service Account", "service account", builderServiceAccountName)
+
+		registryAuthGetter = bcm.authFactory.NewServiceAccountRegistryAuthGetter(bcm.coreClient, mod.Namespace, builderServiceAccountName)
+	}
+
+	imageAvailable, err := bcm.registry.ImageExists(ctx, m.ContainerImage, config.Pull, registryAuthGetter)
+	if err != nil {
+		return build.Result{}, fmt.Errorf("could not check if the image is available: %v", err)
+	}
+
+	if imageAvailable {
+		return build.Result{Status: build.StatusCompleted, Requeue: false}, nil
+	}
+
+	logger.Info("Image not pull-able; building in-cluster")
+
+	buildConfig, err := bcm.ocpBuildsHelper.GetBuildConfig(ctx, mod, targetKernel)
+	if err != nil {
+		if !errors.Is(err, ErrNoMatchingBuildConfig) {
+			return build.Result{}, fmt.Errorf("error getting the build: %v", err)
+		}
+
+		logger.Info("Creating BuildConfig")
+
+		buildConfig, err = bcm.maker.MakeBuildConfig(mod, config, targetKernel, m.ContainerImage)
+		if err != nil {
+			return build.Result{}, fmt.Errorf("could not make BuildConfig: %v", err)
+		}
+
+		if err = bcm.client.Create(ctx, buildConfig); err != nil {
+			return build.Result{}, fmt.Errorf("could not create Job: %v", err)
+		}
+
+		return build.Result{Status: build.StatusCreated, Requeue: true}, nil
+	}
+
+	b, err := bcm.ocpBuildsHelper.GetLatestBuild(ctx, mod.Namespace, buildConfig.Name)
+	if err != nil {
+		return build.Result{}, fmt.Errorf("could not find the latest build: %v", err)
+	}
+
+	switch b.Status.Phase {
+	case buildv1.BuildPhaseComplete:
+		return build.Result{Status: build.StatusCompleted}, nil
+	case buildv1.BuildPhaseNew, buildv1.BuildPhasePending, buildv1.BuildPhaseRunning:
+		return build.Result{Status: build.StatusInProgress, Requeue: true}, nil
+	case buildv1.BuildPhaseFailed:
+		return build.Result{}, fmt.Errorf("buildConfig failed: %v", err)
+	default:
+		return build.Result{}, fmt.Errorf("unknown status: %v", buildConfig.Status)
+	}
+}
+
+//go:generate mockgen -source=manager.go -package=buildconfig -destination=mock_manager.go
+
+type OpenShiftBuildsHelper interface {
+	GetBuildConfig(ctx context.Context, mod kmmv1beta1.Module, targetKernel string) (*buildv1.BuildConfig, error)
+	GetLatestBuild(ctx context.Context, namespace, buildConfigName string) (*buildv1.Build, error)
+}
+
+type openShiftBuildsHelper struct {
```

Comment:
Why do we need this interface?

Reply:
Same here.

Last time we talked about it the conclusions were:
* We shouldn't create new interfaces unless it functionality is being used in another package
* If it functionality is big enough, and we want to test it separately and mock it, it should be extracted to another package, tested separately, and imported by a consumer.

Reply:
> Private interface

Not sure what you mean by "Private interface". But I would create a new package (maybe in the `buildconfig` package - something like `buildconfig.helper` package).

That package will hold the `OpenshiftBuildsHelper` interface and you consume it in `buildconfig` - all the rest, including tests, should remain the same.

Reply:
i am not sure if we need to multiply the packages, but maybe move this interface to a different file under the same package?

Reply:
But then there we still export an interface which is only used internally in the package.

Reply:
A non exported interface (maybe this is what @qbarrand meant by "private interface") is also an option but then how will you test it? Are you OK with testing private interfaces? Also I am not sure we will be able to mock private interfaces.

Reply:
i am not sure if there is such thing as private interfaces... What does it mean? That it is not public?  I think we will be able to test it if the we define unit test in the same package, as we usually do.

Reply:
I don't mind your approach either, I tried to find a win-win solution.
In that case it might be easy to test because those private functions have very lean logic and underlying API k8s calls that can be mock so it that specific case, private function should be OK as well.

In other cases, when the logic isn't very lean, I think we should extract to another interface and export, otherwise, tests are a nightmare.